File: packages/react-native-carplay/android/src/main/java/org/birkir/carplay/CarPlayService.java
package org.birkir.carplay;

import android.content.Intent;

import androidx.annotation.NonNull;

import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import org.birkir.carplay.screens.HelloWorldScreen;

import androidx.car.app.CarAppService;
import androidx.car.app.Screen;
import androidx.car.app.Session;
import androidx.car.app.validation.HostValidator;

public final class CarPlayService extends CarAppService {
  private ReactInstanceManager mReactInstanceManager;
  private Screen screen;

  @Override
  public void onCreate() {
    super.onCreate();
    mReactInstanceManager = ((ReactApplication) getApplication()).getReactNativeHost().getReactInstanceManager();
  }

  @NonNull
  @Override
  public HostValidator createHostValidator() {
    return HostValidator.ALLOW_ALL_HOSTS_VALIDATOR;
  }

  @NonNull
  @Override
  public Session onCreateSession() {
    return new Session() {
      @NonNull
      @Override
      public Screen onCreateScreen(@NonNull Intent intent) {
        return new HelloWorldScreen(getCarContext());
      }
    };
  }
}

File: packages/react-native-carplay/android/src/main/java/org/birkir/carplay/CarPlayService.kt
package org.birkir.carplay

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.util.Log
import androidx.car.app.CarAppService
import androidx.car.app.Session
import androidx.car.app.SessionInfo
import androidx.car.app.validation.HostValidator
import com.facebook.react.ReactApplication
import com.facebook.react.ReactInstanceManager

class CarPlayService : CarAppService() {
  private lateinit var reactInstanceManager: ReactInstanceManager
  override fun onCreate() {
    super.onCreate()
    reactInstanceManager =
      (application as ReactApplication).reactNativeHost.reactInstanceManager
  }

  override fun createHostValidator(): HostValidator {
    return HostValidator.ALLOW_ALL_HOSTS_VALIDATOR
  }

  override fun onCreateSession(sessionInfo: SessionInfo): Session {
    Log.d(TAG, "onCreateSession: sessionId = ${sessionInfo.sessionId}, display = ${sessionInfo.displayType}")
    return CarPlaySession(reactInstanceManager)
  }

  companion object {
    var TAG = "CarPlayService"
  }
}
